Fix issues with setting date/time

chriadam opened the pull request.

chriadam commented:
Not yet ready for review, I still need to implement a simple date picker.

chriadam commented:
Now ready for review.  I based the DateSelector off the TimeSelector, basically, but had to massage stuff so it fit on 5-inch form factor.

jpetrell commented on src/cpuinfo.cpp:
```diff
@@ -24,14 +25,14 @@ int get_cpu_usage(unsigned long long &previousBusy, unsigned long long &previous
 	previousIdle = idleTotal;
 	return qRound((busy * 100.0) / (busy + idle));
 }
+#endif
```
Sorry for the regression!

chriadam replied:
All good, simple fix :-)

jpetrell commented:
Why should the user care about UTC time or the browser local time? These are GX device settings, the time that matters is the GX device time. We shouldn't complicate the UX of the common cases (change the time from GX device, change the time within the same time zone) due to relatively uncommon edge case. From using other devices the user expects to see one time and date, specified in the selected time zone.

Can we have a separate setting for time and date? The combined dialog becomes quite unwieldy.

![image](https://github.com/victronenergy/gui-v2/assets/2203667/d5f22068-681a-47b2-a3f3-0cef32830c33)

If the time zone conversions are difficult can we just provide a setter API over MQTT to set the hour, minute, day, month and year and let the device side do the time setting? Everything is represented in device time, and user just modifies bunch of integers. If that is not possible would just make the date and time setting read-only in the problematic case (accessed from the browser + time zones don't match).

![image](https://github.com/victronenergy/gui-v2/assets/2203667/c31098fb-efd9-4273-a05e-335b109d4449)

DanielMcInnes commented:
I tried this on a cerbo, and it segfaulted the first time I changed the time. Haven't been able to reproduce it. Also, the buttons on the date / time pickerare too sensitive, often when I try to press a button once, it increments multiple times

chriadam commented:
> Why should the user care about UTC time or the browser local time? These are GX device settings, the time that matters is the GX device time. We shouldn't complicate the UX of the common cases (change the time from GX device, change the time within the same time zone) due to relatively uncommon edge case. From using other devices the user expects to see one time and date, specified in the selected time zone.

I agree - but unfortunately we cannot (in WASM builds) convert any date-time to the CerboGX device's timezone, as Qt For WebAssembly doesn't support arbitrary timezones.  The only supported timezones are UTC and the Browser's local timezone.

> Can we have a separate setting for time and date? The combined dialog becomes quite unwieldy.

Hmm, ok, I can prototype that.

> ![image](https://private-user-images.githubusercontent.com/2203667/322509960-d5f22068-681a-47b2-a3f3-0cef32830c33.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMyMzA0OTQsIm5iZiI6MTcxMzIzMDE5NCwicGF0aCI6Ii8yMjAzNjY3LzMyMjUwOTk2MC1kNWYyMjA2OC02ODFhLTQ3YjItYTNmMy0wY2VmMzI4MzBjMzMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MTZUMDExNjM0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWUzN2UyMDVkYTNlYTMxNzJhNjRhYjU4Y2QxOTA4OGQ1MTg1MjNhMTY4ZDVlMjRjNDQ3OTE4YThjMTM3N2IzNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.nT5Sd_2fH60BagMsoXR7XkW7AR3Vc_KNeikICaPsQ-w)
> 
> If the time zone conversions are difficult can we just provide a setter API over MQTT to set the hour, minute, day, month and year and let the device side do the time setting? Everything is represented in device time, and user just modifies bunch of integers. If that is not possible would just make the date and time setting read-only in the problematic case (accessed from the browser + time zones don't match).

The timezone conversions are difficult (in the WebAssembly case only).  It's a possibility, I suppose, to expose year/month/day/hour/minute values on MQTT, and just allow the user to set those directly.  @ReinvdZee do you have thoughts about this?  The alternative is, yes, we could make the setting read-only for WebAssembly case (and in fact just hide the "local time" list entry altogether, since we can only display the browser-local time and not the device local time; so all that would remain is the utc time and the device's timezone).

> ![image](https://private-user-images.githubusercontent.com/2203667/322503158-c31098fb-efd9-4273-a05e-335b109d4449.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMyMzA0OTQsIm5iZiI6MTcxMzIzMDE5NCwicGF0aCI6Ii8yMjAzNjY3LzMyMjUwMzE1OC1jMzEwOThmYi1lZmQ5LTQyNzMtYTA1ZS0zMzViMTA5ZDQ0NDkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MTZUMDExNjM0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTEwZGVkY2JhODZlMGQxM2Q1Y2M5MGNlYTYxYzQ4OGZjMGFlNjA3OTBkYzFhYmQ2MmM5NGIxMzZiMWUxMzYzZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.eRtdCf5_2luN2korINg29vg0A20OOuAWobxrdEx57cI)

chriadam commented:
> I tried this on a cerbo, and it segfaulted the first time I changed the time. IIRC, I changed the hours from 22 -> 23. FW v3.40~2. Difficult to reproduce, but I managed to do it by pressing 'Set' immediately after changing the time. Also, the buttons on the date / time picker are too sensitive, often when I try to press a button once, it increments multiple times.
> 
> Update: I haven't been able to reproduce with the latest FW version v3.40~6

Interesting, I will try to repro the crash.  But it might go away altogether if I split the dialogs into date / time separate pickers.

Regarding spinbox button sensitivity - that's not something I will look into in this PR - if it's broken in this PR then it's broken for every use of SpinBox in our UI, and needs to be fixed separately, I think.

chriadam commented:
Ok, ready for re-review.  Two separate buttons (one to set date, one to set time).  Simplified a bunch of code also.  And fixed a bug in the timezone handling.

jpetrell commented on pages/settings/PageTzInfo.qml:
```diff
@@ -30,51 +30,82 @@ Page {
+		interval: 10000
+		repeat: true
+		triggeredOnStart: true
+		running: BackendConnection.applicationVisible
```
Could also use page.animationEnabled

jpetrell replied:
Generally would try to avoid UI timers when the page is not visible, but in practice a timer that goes of once in 10 seconds has very little impact since AFAICS device never suspends.

jpetrell commented on pages/settings/PageTzInfo.qml:
```diff
@@ -30,51 +30,82 @@ Page {
+						root._openDateSelector()
+					} else {
+						root._openTimeSelector()
+					}
```
Personally would prefer (and find more readable) to declare the buttons explicitly i.e. something like

```
ListItem {
    content: Row {
        Button {
            text: ClockTime.currentDate
            onClicked: root._openDateSelector()
        }
        Button {
            text: ClockTime.currentTime
            onClicked: root._openTimeSelector()
        }
    }
}
```

chriadam replied:
Sure, can do.

jpetrell replied:
Nice! :pray:

chriadam commented on src/clocktime.h:
```diff
@@ -23,44 +23,84 @@ class ClockTime : public QObject
+	Q_PROPERTY(int hour READ hour NOTIFY hourChanged)
+	Q_PROPERTY(int minute READ minute NOTIFY minuteChanged)
+	Q_PROPERTY(int second READ second NOTIFY secondChanged)
+	Q_PROPERTY(int msec READ msec NOTIFY msecChanged)
```
True, should not be necessary any more, will remove.

chriadam replied:
More testing shows that I was wrong - they are still needed.

Setting the timezone setting on the device sets the dbus/mqtt setting, but venus-platform does not change the actual system timezone setting (on the office cerbo, the timezone reported by `date` always returns `UTC`.  As such, the JS engine in QML always does QDateTime->Date conversion which involves a timezone shift.

I need to expose the individual parts.

chriadam commented on pages/settings/PageTzInfo.qml:
```diff
@@ -67,20 +64,28 @@ Page {
+					+ " " + Utils.pad(ClockTime.currentDateTime.hour, 2)
+					+ ":" + Utils.pad(ClockTime.currentDateTime.minute, 2)
 				writeAccessLevel: VenusOS.User_AccessType_User
 				enabled: Global.systemSettings.time.isValid
```
Mock data exists, it does: `ClockTime.clockTime = (new Date().getTime() / 1000)` which should set the appropriate setting, I believe.

DanielMcInnes replied:
If I run on the desktop, the 'set Data / Time' button is disabled
![Screenshot from 2024-04-17 07-31-54](https://github.com/victronenergy/gui-v2/assets/6416318/257a0913-dfbb-4f76-a078-00c7662481f7)

chriadam replied:
You're right!  I was testing on desktop with `--mqtt <cerbo_ip>` not `--mock` before, but yes the mock system settings impl never set the mock MQTT value.  Fixed now.

jpetrell commented:
Serj made following design:

![image](https://github.com/victronenergy/gui-v2/assets/2203667/f51e35a3-511d-4423-8911-e345f4950fa4)

chriadam commented:
> Serj made following design (#930 has SectionHeader component for "UTC" and "Local"):
> 
> ![image](https://private-user-images.githubusercontent.com/2203667/322812705-f51e35a3-511d-4423-8911-e345f4950fa4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzMjQwMzUsIm5iZiI6MTcxMzMyMzczNSwicGF0aCI6Ii8yMjAzNjY3LzMyMjgxMjcwNS1mNTFlMzVhMy01MTFkLTQ0MjMtODkxMS1lMzQ1ZjQ5NTBmYTQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MTdUMDMxNTM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2I3NzZiYzMyYTEwNzYzNWM3NDA5MjdlYjZkMThkNDMxNDNlYzEyNGIwY2Y2NGYxNTNkNzM3OWQwYmMzYjA0MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.VfFXxExv-OaQVTt1wndgAEN6hEx3TQ46BGu7lhbqYjs)

I really don't think we should be accepting design changes for 1.0.0 at this point.

chriadam commented:
Added back the int accessors to avoid having to return a QDateTime to QML (i.e. JS Date conversion with associated timezone shift), and also fixed a check in the setDateTime mutator.

blammit commented on pages/settings/PageTzInfo.qml:
```diff
@@ -67,11 +67,16 @@ Page {
 			}
 
```
Noticed this second commit says "UTC to ?TZ" -> a stray question mark?

chriadam replied:
No, it's meant to suggest that the target timezone can be anything.

blammit commented on components/dialogs/DateSelectorDialog.qml:
```diff
@@ -9,85 +9,20 @@ import Victron.VenusOS
 ModalDialog {
 	id: root
 
-	property var date: ClockTime.currentDateTime
```
The previous DateSelectorDialog design was from Serj -- is there a particular reason for changing this? (Personally I prefer the current one with larger text and buttons, which is more consistent with other dialogs where we use spinboxes.) Do we really need to do both the date and time setting in the same dialog?

Also this commit message says DateSelectorDialog is not used, but it is currently used by ListDateSelector, which is used by PageGeneratorTestRun. Opening "Settings > Generator start/stop > Settings > Conditions > Periodic Run > Run interval start date" now gives this error:

`qrc:/qt/qml/Victron/VenusOS/components/listitems/ListDateSelector.qml: Setting initial properties failed: DateSelectorDialog does not have a property called date`

chriadam replied:
Date and time settings aren't in the same dialog, though.
Oh, I thought the previous one was only a placeholder, not a proper design - the reason I believed so is that its spinboxes are vertically oriented unlike the horizontal time ones.  It was jarring to go from horizontal time to vertical date.

chriadam replied:
Basically reverted my changes to the DateSelectorDialog (at least, as much as I could.  The old DateSelectorDialog API won't work properly for our purposes since its API includes a JS Date, which necessitates a potential tz shift.  So, I fixed that, then updated that case you identified where it was used.  Visually, it should look the same as the old one.)

blammit replied:
Great, works well!

blammit commented on src/clocktime.cpp:
```diff
@@ -162,6 +162,20 @@ bool ClockTime::isDateValid(int year, int month, int day) const
+		return 29;
+	} else {
+		return 28;
+	}
```
Could this just use `calendar.daysInMonth()`?

chriadam replied:
yes, somehow I missed that.  will update.

blammit commented on components/DateSelector.qml:
```diff
@@ -0,0 +1,81 @@
+			to: 2100
+			label.text: value
+			//% "YYYY"
+			//~ A label identifying the "year" portion of a date string
```
I've been using `//:` for translation hints in this codebase.

chriadam replied:
my mistake, will fix

jpetrell commented on pages/settings/PageTzInfo.qml:
```diff
@@ -30,51 +30,85 @@ Page {
-				button.text: ClockTime.currentTimeText
 				writeAccessLevel: VenusOS.User_AccessType_User
-				enabled: Global.systemSettings.time.isValid
+				enabled: allowed && Global.systemSettings.time.isValid
```
allowed=false means invisible, no need to disable

jpetrell commented on pages/settings/PageTzInfo.qml:
```diff
@@ -30,51 +30,85 @@ Page {
+				// Qt for WebAssembly doesn't support timezones,
+				// so we can't display the device-local date/time,
+				// as we don't know what it is.  Just hide the setting.
+				allowed: Qt.platform.os != "wasm"
```
should this have "&& defaultAllowed" (aka userHasReadAccess)?

chriadam replied:
yes, good catch, will fix this.

chriadam replied:
done.

jpetrell reviewed (approved):
Looks good. Didn't find issues in testing. :ok_hand:

chriadam commented:
Thanks, merging.
